core/trino-main/src/main/java/io/trino/sql/planner/EqualityInference.java

```diff
@@ -272,20 +274,19 @@ public static EqualityInference newInstance(Metadata metadata, Collection<Expres
 
     private Expression rewrite(Expression expression, Predicate<Symbol> symbolScope, boolean allowFullReplacement)
     {
-        Set<Expression> subExpressions = SubExpressionExtractor.extract(expression);
+        Stream<Expression> subExpressions = SubExpressionExtractor.extract(expression);
         if (!allowFullReplacement) {
-            subExpressions = subExpressions.stream()
-                    .filter(e -> !e.equals(expression))
-                    .collect(Collectors.toSet());
+            subExpressions = subExpressions
+                    .filter(e -> !e.equals(expression));
```

Comment:
You can do this inline
```java
SubExpressionExtractor.extract(expression)
        .filter(allowFullReplacement
                ? subExpression -> true
                : subExpression -> !subExpression.equals(expression))
        .distinct()
        .forEach(subExpression -> {
            // ...
        });
```

this will allow you to inline `subExpressions` variable

(generally, Streams are single-use, so it's usually better to _not_ have them on a variable)